### ISSUE TITLE
DL3018 false positive when running with set -ex; apk add [...]

### ISSUE THREAD

**Issue by iperdomo** [hadolint/hadolint#196]
- [x] This is a bug report
- [ ] This is a feature request
- [x] I searched existing issues before opening this one

### Expected behavior

`hadolint` reports no errors

### Actual behavior

`hadolint` reports:

```bash
/dev/stdin:3 DL3018 Pin versions in apk add. Instead of `apk add <package>` use `apk add <package>=<version>`
```

### Steps to reproduce the behavior

* Create a `Dockerfile` with a `RUN` directive using `apk` package manager
* Prefix the apk call with a `set -ex;`
* Use `hadolint` as Docker container

### hadolint version
`docker run --rm hadolint/hadolint hadolint --version`: Haskell Dockerfile Linter v1.6.2-0-g76324aa

**Dockerfile** (if relevant)

```Dockerfile
FROM alpine:3.7

RUN set -ex; apk add --no-cache ca-certificates=20171114-r0
```

**Additional environment details (OS, stack version, etc.)**
* Arch Linux - Linux 4.16.4-1-ARCH

````
$ docker version
Client:
 Version:	18.04.0-ce
 API version:	1.37
 Go version:	go1.10.1
 Git commit:	3d479c0af6
 Built:	Tue Apr 17 22:46:17 2018
 OS/Arch:	linux/amd64
 Experimental:	false
 Orchestrator:	swarm

Server:
 Engine:
  Version:	18.04.0-ce
  API version:	1.37 (minimum version 1.12)
  Go version:	go1.10.1
  Git commit:	3d479c0af6
  Built:	Tue Apr 17 22:46:48 2018
  OS/Arch:	linux/amd64
  Experimental:	false
````

**Comment by jbergstroem** [comment]
I don't think this is related to `set`; rather passing `--no-cache`. This explodes for me:

```console
$ hadolint Dockerfile
Dockerfile:10 DL3018 Pin versions in apk add. Instead of `apk add <package>` use `apk add <package>=<version>`
$ tail -n2 Dockerfile
RUN apk add --no-cache -u libtasn1=4.12-r3
ENTRYPOINT exec java $JAVA_OPTS -jar /typeahead-api.jar
```

**Comment by iperdomo** [comment]
This Dockerfile validates just fine
````
FROM alpine:3.7

RUN apk add --no-cache curl=7.59.0-r0
````
And this one fails with the same error
````
FROM alpine:3.7

RUN set -ex; apk add --no-cache curl=7.59.0-r0
````
Output:
```
/dev/stdin:3 DL3018 Pin versions in apk add. Instead of `apk add <package>` use `apk add <package>=<version>`
```
So I'm not sure if is related to `--no-cache`

**Comment by zemanlx** [comment]
@iperdomo 
tl;dr: your issue is that there is no space before `;`

long version: we take a string in shell form and convert it to list of strings where delimiter is **whitespace** https://github.com/hadolint/language-docker/blob/master/src/Language/Docker/Parser.hs#L352 and then split commands in https://github.com/hadolint/hadolint/blob/master/src/Hadolint/Rules.hs#L173 so we can do check for each commands separatelly. 

parsing without space
```
$ stack repl
> parseString "RUN set -ex; apk add --no-cache ca-certificates=20171114-r0"
Right [InstructionPos {instruction = Run (Arguments ["set","-ex;","apk","add","--no-cache","ca-certificates=20171114-r0"]), sourcename = "<string>", lineNumber = 1}]
> bashCommands ["set","-ex;","apk","add","--no-cache","ca-certificates=20171114-r0"]
[["set","-ex;","apk","add","--no-cache","ca-certificates=20171114-r0"]]
```
parsing with space
```
$ stack repl
> parseString "RUN set -ex ; apk add --no-cache ca-certificates=20171114-r0"
Right [InstructionPos {instruction = Run (Arguments ["set","-ex",";","apk","add","--no-cache","ca-certificates=20171114-r0"]), sourcename = "<string>", lineNumber = 1}]
> bashCommands ["set","-ex",";","apk","add","--no-cache","ca-certificates=20171114-r0"]
[["set","-ex"],["apk","add","--no-cache","ca-certificates=20171114-r0"]]
```
This issue should be fixed in https://github.com/hadolint/language-docker

@jbergstroem your one is failing because ` -u, --upgrade           Prefer to upgrade package` is not supported option in Docker best practices.

**Comment by jbergstroem** [comment]
> @zelmanx said your one is failing because `-u, --upgrade Prefer to upgrade package`

Thanks for elaborating!

**Comment by iperdomo** [comment]
@zemanlx thanks. i was using a month old docker image.

I tested it with the most recent one:

```
$ docker run --rm -i hadolint/hadolint:latest hadolint -v
Haskell Dockerfile Linter v1.6.5-0-g933c2e2

$ docker run --rm -i hadolint/hadolint:latest < Dockerfile
/dev/stdin:3 DL3018 Pin versions in apk add. Instead of `apk add <package>` use `apk add <package>=<version>`
````
So, I guess the parsing fix is not yet published?

Again, thanks for the support

**Comment by zemanlx** [comment]
@iperdomo sorry for my confusing comment. I said `it should be fixed` in meaning `somebody should fix it in future`. I have raised an issue in _hadolint/language-docker_.

**Comment by egel** [comment]
The repo with the list of available packages for each alpine version was the most helpful for me.

https://pkgs.alpinelinux.org/packages
